### PR TITLE
build: Drop bitcoin-tx and bitcoin-wallet dependencies on libevent

### PR DESCRIPTION
**Pull request by ryanofsky**
This fixes compile errors trying to build bitcoin-tx and bitcoin-wallet without libevent, which were reported by Luke Dashjr in https://github.com/bitcoin/bitcoin/issues/18465

The fix avoiding `bitcoin-tx` dependency on libevent just adds a conditional build rule. This is implemented in the first commit (more details in commit description).

The fix avoiding `bitcoin-wallet` dependency on libevent requires minor code changes, because `bitcoin-wallet` (unlike `bitcoin-tx`) links against code that calls `urlDecode` / `evhttp_uridecode`. This fix is implemented in the second commit (again details in the commit description).

### CODE REVIEW

**Comment by ryanofsky** [comment]
Updated ddfe37302dcbf93e031704d5eb8af3c508f71f85 -> 01a3392b1b778fa4fcf568013326d6ea1de4fb3b ([`pr/dep-libevent.1`](https://github.com/ryanofsky/bitcoin/commits/pr/dep-libevent.1) -> [`pr/dep-libevent.2`](https://github.com/ryanofsky/bitcoin/commits/pr/dep-libevent.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/dep-libevent.1..pr/dep-libevent.2)) with msvc fix

**Comment by jonasschnelli** [comment]
utACK 01a3392b1b778fa4fcf568013326d6ea1de4fb3b.
Gitian Build: https://bitcoin.jonasschnelli.ch/gitian/build/74

**Comment by MarcoFalke** [comment]
cc @luke-jr

**Review by luke-jr — APPROVED** [review]
Kind of ugly, but utACK

Nit: `URL_DECODE` should probably be named `g_url_decode_fn` or something...

**Comment by ryanofsky** [comment]
> Nit: `URL_DECODE` should probably be named `g_url_decode_fn` or something...

Constant naming is used because it is a constant, not a global variable

**Comment by MarcoFalke** [comment]
> Constant naming is used because it is a constant, not a global variable

`G_URL_DECODE` like `G_TRANSLATION_FUN` :grimacing: ?

**Comment by luke-jr** [comment]
`G_URL_DECODE_FUN` sounds better in that case...

But it's not really a constant - its value varies between builds.

**Comment by MarcoFalke** [comment]
The docs say "compile time constants", not "build time constants"

**Comment by MarcoFalke** [comment]
Anyway, I was mostly joking about that style nit. Let's not hold back this pull based on naming guidelines.

**Comment by luke-jr** [comment]
It's definitely not a compile time constant! ;)

**Comment by ryanofsky** [comment]
> It's definitely not a compile time constant! ;)

FWIW, following this up in #18568

**Comment by practicalswift** [comment]
Concept ACK, but wouldn't it be easier to implement `std::string urlDecode(const std::string &urlEncoded)` without `evhttp_uridecode`?

It has always seemed weird to me that we depend on third-party code for such a trivial function :)

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -77,9 +77,9 @@ bool HaveKey(const SigningProvider& wallet, const CKey& key)
 
 bool GetWalletNameFromJSONRPCRequest(const JSONRPCRequest& request, std::string& wallet_name)
 {
-    if (request.URI.substr(0, WALLET_ENDPOINT_BASE.size()) == WALLET_ENDPOINT_BASE) {
+    if (URL_DECODE && request.URI.substr(0, WALLET_ENDPOINT_BASE.size()) == WALLET_ENDPOINT_BASE) {
```
Comment (sidhujag) [inline src/wallet/rpcwallet.cpp:80]: Shouldn't this be #ifdef URL_DECODE, on bitcoin-tx the symbol does not exist if libevent is not used.

Comment (MarcoFalke) [reply src/wallet/rpcwallet.cpp:80]: The symbol `URL_DECODE` is always defined

Comment (sidhujag) [reply src/wallet/rpcwallet.cpp:80]: my bad, thx

**Pull request merged**